src/Week7/HW7.hs

```diff
@@ -0,0 +1,240 @@
+getElts :: [Int] -> Vector a -> Maybe [a]
+getElts lst v = mapM' (v V.!?) lst
+
+-- Exercise 3 -----------------------------------------
+
+type Rnd a = Rand StdGen a
+
+-- Use the randomness monad to produce a random element of a vector
+randomElt :: Vector a -> Rnd (Maybe a)
+randomElt v = case length v of
+  0 -> return Nothing
+  l -> (v V.!?) <$> getRandomR (0, l-1)
+
+-- Exercise 4 -----------------------------------------
+-- Write a function that outputs a vector of specified length with random elements
+randomVec :: Random a => Int -> Rnd (Vector a)
+randomVec l = sequence $ V.generate l (const getRandom)
+
+-- Same as above, but with elements within a range
+randomVecR :: Random a => Int -> (a, a) -> Rnd (Vector a)
+randomVecR l (a,b) = sequence $ V.generate l (const $ getRandomR (a,b))
+
+-- Exercise 5 -----------------------------------------
+--  Implement the Fisher-Yates algorithm to shuffle the elements of a vector
+
+swapV' :: Int -> Int -> Vector a -> Vector a
+swapV' a b v = Ma.fromJust $ swapV a b v
+--swapV' = Ma.fromJust . swapV 
+-- why can we not leave out the parameters, as in the line above?
+
+shuffle :: Vector a -> Rnd (Vector a)
+shuffle vt = helper (pure vt) (length vt - 1)
+  where helper v 0 = v
+        helper v n = helper (swapV' n <$> getRandomR (0, n-1) <*> v) (n - 1)
+
+-- Exercise 6 -----------------------------------------
+-- Implement a function that partitions a vector around the element at a given index
+
+-- Appending an element to a vector is O(N), so we first convert the input to a list (to use O(1) cons) and then convert the result back to a vector at the end
+partitionAt :: Ord a => Vector a -> Int -> (Vector a, a, Vector a)
+partitionAt v i = toTupleV $ helper [] (v!i) [] i (V.toList v)
+  where helper lt p gt _ []     = (lt, p, gt)
+        helper lt p gt 0 (_:xs) = helper lt p gt (-1) xs --skip the pivot
+        helper lt p gt n (x:xs) = if x < p then helper (x:lt) p gt (n-1) xs
+                                         else helper lt p (x:gt) (n-1) xs
+
+toTupleV :: ([a], a, [a]) -> (Vector a, a, Vector a)
+toTupleV (fs, i, sn) = (V.fromList fs, i, V.fromList sn)
+
+
+-- Exercise 7 -----------------------------------------
+-- Define a naive quicksort for vectors
+
+-- Quicksort (given to us)
+quicksort :: Ord a => [a] -> [a]
+quicksort [] = []
+quicksort (x:xs) = quicksort [ y | y <- xs, y < x ]
+                   <> (x : quicksort [ y | y <- xs, y >= x ])
+
+
+qsort :: Ord a => Vector a -> Vector a
+qsort v = V.fromList $ quicksort $ V.toList v
+
+-- or, using vectors directly (inefficient)
+qsort' :: Ord a => Vector a -> Vector a
+qsort' v = case V.uncons v of
+   Nothing     -> V.empty
+   Just (a,as) -> qsort' [x | x <- as, x < a]
+                  <> V.cons a (qsort' [x | x <- as, x >= a])
+-- Is there a way to use LambdaCase when we are pattern matching a function
```

Comment:
There is [ViewPatterns](https://downloads.haskell.org/ghc/8.8.3/docs/html/users_guide/glasgow_exts.html#extension-ViewPatterns) language extension that would make something like this possible.